crates/binjs_io/src/entropy/predict.rs

```diff
@@ -336,3 +346,172 @@ impl<NodeValue> PathPredict<NodeValue, SymbolInfo> where NodeValue: Eq + Hash +
     }
 }
 
+
+
+
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+enum WindowPrediction {
+    BackReference(BackReference),
+    DictionaryIndex(DictionaryIndex),
+}
+
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    /// The window width
+    width: usize,
+
+    /// All the known values.
+    value_by_dictionary_index: Vec<NodeValue>,
+    dictionary_index_by_value: HashMap<NodeValue, DictionaryIndex>,
+
+    latest_values: Vec<NodeValue>,
+
+    info: ContextInformation<WindowPrediction, Statistics>,
+}
+impl<NodeValue, Statistics> WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    pub fn new(width: usize) -> Self {
+        WindowPredict {
+            width,
+            value_by_dictionary_index: Vec::with_capacity(1024), // FIXME: Magic number
+            dictionary_index_by_value: HashMap::with_capacity(1024),
+            latest_values: Vec::with_capacity(width),
+            info: ContextInformation::new(),
+        }
+    }
+
+    /// Update current window by moving an index to the first position.
+    fn update_window_by_index(&mut self, index: BackReference) -> Result<BackReference, ()> {
+        let as_usize: usize = index.into();
+        if as_usize >= self.latest_values.len() {
+            return Err(());
+        }
+        let ref mut slice = self.latest_values
+            .as_mut_slice()
+            [0..index.into()];
+        slice.rotate_right(1);
+        Ok(index)
+    }
+
+    /// Update current window by putting a value to the first position.
+    ///
+    /// If the value is already in the window, it is moved to the first position,
+    /// otherwise, it is added. If the resulting window is too large, it is truncated.
+    fn update_window_by_value(&mut self, value: &NodeValue) -> Option<BackReference> {
+        // It's possible that the value was present in the window.
+        if let Some(index) = self.latest_values
+            .iter()
+            .position(|v| v == value)
+        {
+            return Some(self.update_window_by_index(BackReference(index))
+                .unwrap())
+        }
+        if self.latest_values.len() < self.width {
+            // If the window isn't full yet, simply add the value at start.
+            self.latest_values.insert(0, value.clone());
+        } else {
+            // Otherwise, rotate.
```

Comment:
"rotate" doesn't correctly explains what happens.
this pushes the value in front of the vector and drops the last value.


btw, if this may happen frequently, can't you use `vec_deque` ?
https://doc.rust-lang.org/std/collections/vec_deque/index.html

Reply:
> btw, if this may happen frequently, can't you use `vec_deque` ?
> https://doc.rust-lang.org/std/collections/vec_deque/index.html

This would improve the performance of `window_insert_value` and reduce that of `window_move_to_front`. I actually have no idea which one is most common, but since both of them are bounded by `width`, which should be fairly small, I don't think it's going to be a major difference.

Let's reconvene when we need to profile :)